packages/drivers/odsp-driver/src/zipItDataRepresentationUtils.ts

```diff
@@ -424,25 +451,61 @@ export class NodeCore {
                 case MarkerCodes.Int32:
                 case MarkerCodes.Int64:
                 {
-                    this.children.push(buffer.read(getValueSafely(codeToBytesMap, code)));
+                    children.push(buffer.read(getValueSafely(codeToBytesMap, code)));
                     continue;
                 }
                 case MarkerCodes.BoolTrue:
-                    this.children.push(true);
+                    children.push(true);
                     continue;
                 case MarkerCodes.BoolFalse:
-                    this.children.push(false);
+                    children.push(false);
                     continue;
                 case MarkerCodesEnd.list:
-                    assert(this.type === "list", "Mismatch in end of list");
-                    return;
                 case MarkerCodesEnd.set:
-                    assert(this.type === "set", "Mismatch in end of set");
-                    return;
+                    // Note: We are not checking that end marker matches start marker.
+                    // I.e. that we do not have a case where we start a 'list' but end with a 'set'
+                    // Checking it would require more state tracking that seems not very useful, given
+                    // our code does not care.
+                    children = stack.pop()!;
+
+                    // To my surprise, checking children !== undefined adds measurable cost!
+                    // We will rely on children.push() crashing in case of mismatch, and check below
+                    // (outside of the loop)
+                    continue;
                 default:
                     throw new Error(`Invalid code: ${code}`);
             }
         }
+
+        assert(children === this.children, "Unpaired start/end list/set markers!");
+
+        /**
+         * Process all the strings at once!
+         */
+        let length = 0;
+        for (const el of strings) {
+            length += el.endPos - el.startPos + 1;
+        }
+        const stringBuffer = new Uint8Array(length);
+
+        length = 0;
+        const input = buffer.buffer;
+        assert(input.byteOffset === 0, "code below assumes no offset");
+
+        for (const el of strings) {
+            for (let it = el.startPos; it < el.endPos; it++) {
+                stringBuffer[length] = input[it];
+                length++;
+            }
+            stringBuffer[length] = 0;
+            length++;
+        }
+        assert(length === stringBuffer.length, "test");
```

Comment:
Q: should I add recovery code here if I learn that there are \0 in the payload?
This is mostly about DDS / DataStore names, and op content. We can enforce that in runtime (catch earlier). Or recover here. Or do nothing - not sure. Chances that strings with \0 would be Ok across stack (including storage) is low

Reply:
Fixed